crates/polars-io/src/json/mod.rs

```diff
@@ -274,27 +274,41 @@ where
                 let mut bytes = rb.deref().to_vec();
                 let json_value =
                     simd_json::to_borrowed_value(&mut bytes).map_err(to_compute_err)?;
-
-                // struct type
-                let dtype = if let Some(mut schema) = self.schema {
+                let dtype_result = if let Some(mut schema) = self.schema {
                     if let Some(overwrite) = self.schema_overwrite {
                         let mut_schema = Arc::make_mut(&mut schema);
                         overwrite_schema(mut_schema, overwrite)?;
                     }
-
-                    DataType::Struct(schema.iter_fields().collect()).to_arrow(CompatLevel::newest())
+                    Ok(DataType::Struct(schema.iter_fields().collect())
+                        .to_arrow(CompatLevel::newest()))
                 } else {
                     // infer
-                    let inner_dtype = if let BorrowedValue::Array(values) = &json_value {
-                        infer::json_values_to_supertype(
+                    let inner_dtype_result = if let BorrowedValue::Array(values) = &json_value {
+                        let supertype = infer::json_values_to_supertype(
                             values,
                             self.infer_schema_len
                                 .unwrap_or(NonZeroUsize::new(usize::MAX).unwrap()),
-                        )?
-                        .to_arrow(CompatLevel::newest())
+                        );
+                        match supertype {
+                            Ok(supertype) => Ok(supertype.to_arrow(CompatLevel::newest())),
+                            Err(e) => Err(e),
+                        }
                     } else {
-                        polars_json::json::infer(&json_value)?
+                        polars_json::json::infer(&json_value)
                     };
+                    if inner_dtype_result.is_err() {
+                        match &json_value {
```

Comment:
Instead of changing the whole `dtype` inference and control flow. Could we start this function with this check, and just early return `DataFrame::empty`?